components/renku_data_services/crc/db.py

```diff
@@ -858,3 +866,72 @@ async def update_user(self, api_user: base_models.APIUser, keycloak_id: str, **k
             if (no_default_access := kwargs.get("no_default_access")) is not None:
                 user.no_default_access = no_default_access
             return user.dump()
+
+
+@dataclass
+class ClusterRepository:
+    """Repository for cluster descriptions."""
+
+    session_maker: Callable[..., AsyncSession]
+
+    @_only_admins
+    async def select_all(self, api_user: base_models.APIUser) -> list[Cluster]:
+        """Get cluster descriptions from the database."""
+
+        async with self.session_maker() as session:
+            r = await session.scalars(select(ClusterORM))
+            clusters = r.fetchall()
+
+            return [c.to_cluster() for c in clusters]
+
+    @_only_admins
+    async def select(self, api_user: base_models.APIUser, cluster_id: ULID) -> Cluster:
+        """Get cluster descriptions from the database."""
+
+        async with self.session_maker() as session:
+            r = await session.scalars(select(ClusterORM).where(ClusterORM.id == cluster_id))
+            cluster = r.one_or_none()
+            if cluster is None:
+                raise errors.MissingResourceError(message=f"Cluster definition id='{cluster_id}' does not exist.")
+
+            return cluster.to_cluster()
+
+    @_only_admins
+    async def insert(self, api_user: base_models.APIUser, cluster: UnsavedCluster) -> Cluster:
+        """Creates a new cluster description."""
+
+        cluster_orm = ClusterORM.from_model(cluster)
+        async with self.session_maker() as session, session.begin():
+            session.add(cluster_orm)
+            await session.flush()
+            await session.refresh(cluster_orm)
+
+            return cluster_orm.to_cluster()
+
+    @_only_admins
+    async def update(self, api_user: base_models.APIUser, cluster: SavedCluster, cluster_id: ULID) -> Cluster:
+        """Updates a cluster description."""
+
+        async with self.session_maker() as session, session.begin():
+            saved_cluster = (await session.scalars(select(ClusterORM).where(ClusterORM.id == cluster_id))).one_or_none()
+            if saved_cluster is None:
+                raise errors.MissingResourceError(message=f"Cluster definition id='{cluster.id}' does not exist.")
+
+            saved_cluster.id = cluster.id
```

Comment:
issue: We should not be changing the primary key.

Reply:
In the update method make the cluster parameter be of type "UnsavedCluster" which does not contain an id.

Reply:
done